Added bytecode to project

SirPaws opened the pull request:
I've finished writing the bytecode format, **this is not ready to be merged** but I wanted open a pull request so people can check if I've forgotten stuff or if I can improve upon it

as I mentioned this isn't ready to be merged, I'll be connecting the bytecode with the language tomorrow but for now, I wanted to show the progress I've made and make an early PR.
also, this next bit is more for the report but I'll write it as a 'note' here.

# design
in reality, I didn't really consider the design of the bytecode I just used what I've seen before, but with that said I was thinking of some of the instruction sets and intermediate representations I've had a look at before 
in this case, I was thinking of the RISC-V/X86 instruction sets, MSIL/CIL[^1] and MIR[^2]

I've included the textual format[^3] that I used as a guide for implementing it (**;** is the start of a comment)

oh and also I haven't had time to implement a binary output format but I've thought about it 
```
HEADER
SYMBOL_TABLE
INSTRUCTIONS ....
```
the header would just be an identifier to make sure it's actually bytecode we're reading, something like 4 bytes saying **CODE** in ascii.
the symbol table is not actual language symbols but identifiers for variables, so it's essentially just a list of null-terminated strings.
the next bit is the instructions these would be read as packets of 13 bytes `MNEMONIC | ARG0 | ARG1 | ARG2` 
- 1 byte mnemonic (instruction or whatever you want to call it)
- 3 * 32-bit unsigned integer representing an offset in the symbol table

the only caveat is that `func` and `call` need special handling, but they are still loaded in the same way
for `func` the first integer is the number of parameters and the second integer will be the number of instructions that this function uses.
the following bytes would then be the parameters (still 32-bit unsigned integers) followed by the instructions (for `func` only)
```
func num_params, num_instructions, 0
parameter_names....
instructions....
```

for call the first integer is the 'return address or the variable where the result is stored, the next is the actual function we want to call, followed by the number of arguments. 
same as with `func` the next bytes would then be the parameters
```
call return_variable, function, num_params
parameter_names...
``` 

[^1]: https://github.com/vnmakarov/mir
[^2]: https://docs.microsoft.com/en-us/dotnet/standard/managed-code
[^3]: the example_bytecode.bt file

SirPaws committed:
Added bytecode project

this is temporary and will be moved into the right directory when it has been connected to the language

miningape reviewed (approved):
Looks good to me, I just wished I knew you were making this then I wouldnt have made my own shitty bytecode :P

I'll use both in my report (quick and dirty vs proper implementation) so can we set up a call next week where I ask you a few questions?

miningape commented:
Also I didnt see any code that explicitely tells me your handling closures, are you?

miningape commented on bytecode/src/execute.py:
```diff
@@ -0,0 +1,231 @@
+        Mnemonics.DIVISION       : division_closure.evaluate,
+        Mnemonics.MULTIPLICATION : multiplication_closure.evaluate,
+    }
+    assert len(execute_byte.keys()) == len(Mnemonics)
```
what is this for?

SirPaws replied:
I wanted something like a switch in python and the assertion is just to make sure it's exhaustive

miningape replied:
Yeah I wanted to use that but there isnt a stable version for apple silicon and I didnt feel like building from source

SirPaws committed:
Update bytecode/src/bytecode.py

Co-authored-by: Kyle Johnson <kyle@weld.app>

SirPaws committed:
Connected bytecode with language

here's a bit of a changelog, it's not an exact order but just to show most of the stuff that has changed
[1] moved bytecode files from temporary folder into actual location
[1] added bytecode generator that takes in a file and generates bytecode
[3] renamed 'bytecode' Literal to Value to not confuse with AST
[2] added missing bytecodes. namely NotEquals, unary operators (!, -, +) and some dealing with scope (Frame, Raze)
[3] edited main.pych to test more functionality
[4] added printing of bytecode
[5] added extra command line arguments to allow generating/executing bytecode, also for printing said bytecode
[6] did some minor cleanup, but more could be used
[7] removed bytecode temporary folder (I kept the example_bytecode.bt file because it shows some of the initial ideas)

SirPaws commented:
okay so the bytecode stuff is now finished (still no binary format) it's fully connected with the language.
I wrote a 'changelog' in the commit description but I'll also post it here so it's easier to see 

1.  moved bytecode files from temporary folder into actual location
2.  added bytecode generator that takes in a file and generates bytecode
3.  renamed 'bytecode' Literal to Value to not confuse with AST
4.  added missing bytecodes. namely NotEquals, unary operators (!, -, +) and some dealing with scope (Frame, Raze)
5.  edited main.pych to test more functionality
6.  added printing of bytecode
7.  added extra command line arguments to allow generating/executing bytecode, also for printing said bytecode
8.  did some minor cleanup, but more could be used
9.  removed bytecode temporary folder (I kept the example_bytecode.bt file because it shows some of the initial ideas)

SirPaws committed:
Merge branch 'main' into bytecode

SirPaws committed:
Merge branch 'main' into bytecode

SirPaws committed:
return and while loops supported in bytecode

SirPaws committed:
arrays supported in bytecode

SirPaws commented:
okay so now all the new features are supported, one small note is that I made arrays very forgiving.
right now the bytecode would see code like this as valid
```js
let a = [1, 2, 3]
a[print] = 5
```
It's honestly trivial to do it properly but I don't feel like it and if someone else want's to do it go ahead

SirPaws committed:
minor bugs fixed

SirPaws committed:
fixed bug with handling of native function call

SirPaws commented:
there's also currently a bug with variables, but it is caused by pythons handling of variables and I am not knowledgable enough to be able to fix it currently, and I honestly can't be bothered

miningape commented on src/pychart/_interpreter/visitors/bytecode_generator.py:
```diff
@@ -0,0 +1,415 @@
+
+from typing import Any, Dict, List, Callable
+
+from src.pychart._interpreter.ast_nodes.statement import *
```
nevermind paws used some tricks that arent exactly type friendly and I dont wanna make changes so i'm not gonna be making the pr

SirPaws committed:
Merge branch 'main' into bytecode

spralja commented:
You can make it a draft in the menu to the left

On Thu, May 5, 2022, 17:31 SirPaws ***@***.***> wrote:

> I've finished writing the bytecode format, *this is not ready to be
> merged* but I wanted open a pull request so people can check if I've
> forgotten stuff or if I can improve upon it
>
> as I mentioned this isn't ready to be merged, I'll be connecting the
> bytecode with the language tomorrow but for now, I wanted to show the
> progress I've made and make an early PR.
> also, this next bit is more for the report but I'll write it as a 'note'
> here.
> design
>
> in reality, I didn't really consider the design of the bytecode I just
> used what I've seen before, but with that said I was thinking of some of
> the instruction sets and intermediate representations I've had a look at
> before
> in this case, I was thinking of the RISC-V/X86 instruction sets, MSIL/CIL1
> <#m_6170312573176373430_user-content-fn-1-f1c8e43dc44d3251b08a381bd4a37ff2>
> and MIR2
> <#m_6170312573176373430_user-content-fn-2-f1c8e43dc44d3251b08a381bd4a37ff2>
>
> I've included the textual format <http://bytecode/example_bytecode.bt>
> that I used as a guide for implementing it (*;* is the start of a comment)
>
> oh and also I haven't had time to implement a binary output format but
> I've thought about it
>
> HEADER
>
> SYMBOL_TABLE
>
> INSTRUCTIONS ....
>
>
> the header would just be an identifier to make sure it's actually bytecode
> we're reading, something like 4 bytes saying *CODE* in ascii.
> the symbol table is not actual language symbols but identifiers for
> variables, so it's essentially just a list of null-terminated strings.
> the next bit is the instructions these would be read as packets of 13
> bytes MNEMONIC | ARG0 | ARG1 | ARG2
>
>    - 1 byte mnemonic (instruction or whatever you want to call it)
>    - 3 * 32-bit unsigned integer representing an offset in the symbol
>    table
>
> the only caveat is that func and call need special handling, but they are
> still loaded in the same way
> for func the first integer is the number of parameters and the second
> integer will be the number of instructions that this function uses.
> the following bytes would then be the parameters (still 32-bit unsigned
> integers) followed by the instructions (for func only)
>
> func num_params, num_instructions, 0
>
> parameter_names....
>
> instructions....
>
>
> for call the first integer is the 'return address or the variable where
> the result is stored, the next is the actual function we want to call,
> followed by the number of arguments.
> same as with func the next bytes would then be the parameters
>
> call return_variable, function, num_params
>
> parameter_names...
>
>
> Footnotes
>
>    1.
>
>    https://github.com/vnmakarov/mir ↩
>    <#m_6170312573176373430_user-content-fnref-1-f1c8e43dc44d3251b08a381bd4a37ff2>
>    2.
>
>    https://docs.microsoft.com/en-us/dotnet/standard/managed-code ↩
>    <#m_6170312573176373430_user-content-fnref-2-f1c8e43dc44d3251b08a381bd4a37ff2>
>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/miningape/pychart/pull/61
> Commit Summary
>
>    - 063578e
>    <https://github.com/miningape/pychart/pull/61/commits/063578ea71f58e9a1e88271f3d393fd87d317887>
>    Added bytecode project
>
> File Changes
>
> (17 files <https://github.com/miningape/pychart/pull/61/files>)
>
>    - *A* bytecode/example_bytecode.bt
>    <https://github.com/miningape/pychart/pull/61/files#diff-6952620c20220950f829a1790f7157bc4e7e368308bac6c5f4f521e98050b835>
>    (59)
>    - *A* bytecode/example_bytecode.bt.bak
>    <https://github.com/miningape/pychart/pull/61/files#diff-ac955afde1e89518ff50fbf8ef25fb715182e30d95e82eac3cd06477c55e2333>
>    (49)
>    - *A* bytecode/main.py
>    <https://github.com/miningape/pychart/pull/61/files#diff-d8cfaf61d8e4ebf47f2f3f10500e48e18b314ce68aafef92c8453c7cfdcb80d0>
>    (11)
>    - *A* bytecode/src/__init__.py
>    <https://github.com/miningape/pychart/pull/61/files#diff-86f27172277e5295f4ad5b09fa38362214a9523b9c399de0c91182b048cb3f89>
>    (0)
>    - *A* bytecode/src/bytecode.py
>    <https://github.com/miningape/pychart/pull/61/files#diff-e77886fdfee3d189ec5d0614de51fd69211e80879e956a42c8ed0b0519a8db0b>
>    (50)
>    - *A* bytecode/src/bytecode_arithmetic.py
>    <https://github.com/miningape/pychart/pull/61/files#diff-50f81d7b9a83d2573f839ac1e6ac43d05e8122ea22da9ed55d2fb4c7fd17cd06>
>    (19)
>    - *A* bytecode/src/bytecode_binary.py
>    <https://github.com/miningape/pychart/pull/61/files#diff-4d88c539bb0eecd99de2ec8643e224621ce6fa5f2925149f4201d53f86ab2fb2>
>    (20)
>    - *A* bytecode/src/bytecode_conditional.py
>    <https://github.com/miningape/pychart/pull/61/files#diff-6a3781002e3da52bfa72cda4f9d5993b13ab2cd2be6609acf969eabaab07243c>
>    (31)
>    - *A* bytecode/src/bytecode_function.py
>    <https://github.com/miningape/pychart/pull/61/files#diff-ce68df969cd2f100b585610dd5558a138198c4e13dbd7054a03307fb23a209a5>
>    (76)
>    - *A* bytecode/src/bytecode_jumps.py
>    <https://github.com/miningape/pychart/pull/61/files#diff-8b7b5e84f450cc609be3f9fddb0a328d044b157e9050618d0d941beaf56fe7bc>
>    (45)
>    - *A* bytecode/src/bytecode_util.py
>    <https://github.com/miningape/pychart/pull/61/files#diff-073c78f3b801855cc7a7a877e73cbe20bb16ce5ed8e6e8eaae95162591c7d967>
>    (65)
>    - *A* bytecode/src/bytecodes.py
>    <https://github.com/miningape/pychart/pull/61/files#diff-c2914092dcf77a8d72c819ee1c3ea8f35027720ecb8031e7b1552f0fe42f5077>
>    (7)
>    - *A* bytecode/src/example.py
>    <https://github.com/miningape/pychart/pull/61/files#diff-f550ab458b39bad91fcb29ece6f926e66646e4bfde9b1fc85616c0b2798fee40>
>    (99)
>    - *A* bytecode/src/execute.py
>    <https://github.com/miningape/pychart/pull/61/files#diff-388dcaff0e1c668d7dc405e3a6d7a3147fe37911fcdbafe969ec79c95b5bc223>
>    (231)
>    - *A* bytecode/tests/__init__.py
>    <https://github.com/miningape/pychart/pull/61/files#diff-6ee80e27d91859c0670173708db69fad5b8cddcd2b51f30f1681c33cb22d7b4a>
>    (0)
>    - *A* bytecode/tests/arithmetic_tests.py
>    <https://github.com/miningape/pychart/pull/61/files#diff-60f5962e77fb511e951df67ab044d57ff5faf0d7eefb1ef0241b98b53384d3aa>
>    (125)
>    - *A* bytecode/tests/conditional_tests.py
>    <https://github.com/miningape/pychart/pull/61/files#diff-bdabf18b3cac0f83efd8d763c040593e17c13178e8bbc4b718d9f921b624c501>
>    (2)
>
> Patch Links:
>
>    - https://github.com/miningape/pychart/pull/61.patch
>    - https://github.com/miningape/pychart/pull/61.diff
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/miningape/pychart/pull/61>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANQOJN257HPND56DOXM56YTVIPSUVANCNFSM5VFOMIQQ>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>
